### PR TITLE
Added a unit test to verify that host header is preserved after probe…

### LINKED ISSUE #71592: Probe httpHeaders "Host" is lost after redirect
Hello there,

I did search around and didn't find a report for this already, I hope it's  not a dupe.

**What happened**:
In an HTTP probe definition (readyness or liveness), the custom host header is not set properly after a redirect occurs.

**What you expected to happen**:
Seeing that #18233 states that redirects are weird but allowed, I would expect the custom headers to be maintained after the redirect occurs.

**How to reproduce it (as minimally and precisely as possible)**:
Setup an HTTP service that expects a specific Host header to be used (using Django here) and would return a 400 if the Host doesn't match (for a more detail view at how Django deals with this [read along](https://docs.djangoproject.com/en/2.1/ref/settings/#allowed-hosts)).

Then setup a probe like:

```
livenessProbe:
    httpGet:
        path: /admin/
        port: http
        httpHeaders:
            - name: Host
              value: the-expected-host.tld
```

You should now see in your logs that the first call to the service succeeded with a 301 (or 3XX) which means the Host header was set properly, but that the call to the redirected url will fail with a 400 which means the Host header was not set anymore and makes the probe raise an error.

**Anything else we need to know?**:
In the above test, setting the "path" to the redirected url works and the probe is green.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.3", GitCommit:"2bba0127d85d5a46ab4b778548be28623b32d0b0", GitTreeState:"clean", BuildDate:"2018-05-21T09:17:39Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.7", GitCommit:"0c38c362511b20a098d7cd855f1314dad92c2780", GitTreeState:"clean", BuildDate:"2018-08-20T09:56:31Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: intel NUC
- OS (e.g. from /etc/os-release): Ubuntu 16.04.5 LTS
- Kernel (e.g. `uname -a`): Linux nuc 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: Minikube
- Others:

/kind bug

### PR DESCRIPTION
**Pull request by ahg-g**
… redirect.

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adds a unit test to capture possible regressions in preserving host header after probe redirect 

**Which issue(s) this PR fixes**:

Fixes #71592

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/probe/http/http_test.go [hunk 1]
```diff
@@ -314,3 +314,39 @@ func TestHTTPProbeChecker_NonLocalRedirects(t *testing.T) {
...
+
+func TestHTTPProbeChecker_HostHeaderPreservedAfterRedirect(t *testing.T) {
+	hostHeader := "www.example.com"
+
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.URL.Path {
+		case "/redirect":
+			http.Redirect(w, r, "/success", http.StatusFound)
+		case "/success":
+			if r.Host != hostHeader {
```
Comment (neolit123) [inline pkg/probe/http/http_test.go]: it seems this check will always pass?
`r.Host` probably will be `127.0.0.1...`

Comment (ahg-g) [reply pkg/probe/http/http_test.go]: oops, I fixed the test and added another case to test probe failures.

note that r.URL.Host is different from r.Host, the former is the host part of the url (which in this case will be 127.0.0.1), the latter is the host header value.

**Review by mattjmcnaughton — APPROVED** [review]
Thanks for adding this test :)

/lgtm
/ok-to-test
/assign @tallclair

File: pkg/probe/http/http_test.go [hunk 2]
```diff
@@ -314,3 +314,51 @@ func TestHTTPProbeChecker_NonLocalRedirects(t *testing.T) {
...
+		expectedResult probe.Result
+	}{
+		"success": {successHostHeader, probe.Success},
+		"fail":    {failHostHeader, probe.Failure},
+	}
+	for desc, test := range testCases {
+		headers := http.Header{}
+		headers.Add("Host", test.hostHeader)
+		t.Run(desc+"local", func(t *testing.T) {
+			prober := New(false)
```
Comment (mattjmcnaughton) [inline pkg/probe/http/http_test.go]: Would you mind naming these variables (i.e. `followNonLocalRedirects = false`)?

I created https://github.com/kubernetes/kubernetes/pull/78031/files to make a similar change for other tests in this repo.

Comment (ahg-g) [reply pkg/probe/http/http_test.go]: Done.

**Comment by thockin** [comment]
I think this looks like what we want to prove.

Thanks!

/approve

**Comment by thockin** [comment]
Thanks!

/lgtm

**Pull request merged**